### PR TITLE
Mepts 674 tx rtt plhiv

### PR DESCRIPTION
**Pull request by cfaife**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+            + "                            e.encounter_type =  ${ARVPharmaciaEncounterType} and "
+            + "                            e.location_id = :location and "
+            + "                            e.voided = 0 and "
+            + "                            e.encounter_datetime <= '"
+            + endDate
+            + "' "
+            + "					INNER JOIN obs o on "
+            + "                            o.encounter_id = e.encounter_id and "
+            + "                            o.concept_id = ${returnVisitDateForArvDrugConcept} and "
+            + "                            o.voided = 0 "
```
Comment (amosh101) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java:228]: Maybe we need a group by clause here.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+            + "                        INNER JOIN encounter e on "
+            + "                            e.patient_id = ficha.patient_id and "
+            + "                            e.encounter_datetime = ficha.encounter_datetime and "
+            + "                            e.encounter_type IN (${adultoSeguimentoEncounterType},${aRVPediatriaSeguimentoEncounterType}) and "
+            + "                            e.location_id = :location and "
+            + "                            e.voided = 0 "
+            + "                        INNER JOIN obs o on "
+            + "                            o.encounter_id = e.encounter_id and "
+            + "                            o.concept_id = ${returnVisitDateConcept} and "
+            + "                            o.voided = 0 "
```
Comment (amosh101) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java:253]: same here

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java (hunk 1 above)
Comment (cfaife) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java:228]: I  can't touch this. it  has been taken `as is` from  the  LFTU query according  with user story

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java (hunk 2 above)
Comment (cfaife) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java:253]: the same here, I  can't touch this. it  has been taken `as is` from  the  LFTU query according  with user story

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java (hunk 1 above)
Comment (amosh101) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java:228]: very well, if results work as expected.

**Review by amosh101 — APPROVED** [review]
Works as expected

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/EriDSDCohortQueries.java [hunk 3]
```diff
@@ -447,8 +446,18 @@ public CohortDefinition getPatientsWithViralLoadLessThan1000Within12Months() {
             + "                OR "
             + "                (o.concept_id=${hivViralLoadQualitative} AND o.value_coded IN (${beyondDetectableLimit},${undetectableViralLoad},${lessThan10Copies},${lessThan20Copies},${lessThan40Copies},${lessThan400Copies},${lessThan839Copies})) "
             + "            )  "
-            + "        AND e.location_id= :location  "
-       
+            + "        AND e.location_id= :location "
+            + "AND (  "
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/EriDSDCohortQueries.java]: Were changes on this file required? or mvn clean install will iron those out

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxMlQueries.java [hunk 4]
```diff
@@ -534,7 +534,7 @@ public static String getPatientsWithVisitCardAndWithObs(
             + "                                                    AND        o.voided = 0"
             + "                                                    AND        e.encounter_type = ${aRVPharmaciaEncounterType}"
             + "                                                    AND        e.location_id = :location"
-            + "                                                    AND        e.encounter_datetime <= :onOrBefore"
+            + "                                                    AND        e.encounter_datetime <= :endDate"
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxMlQueries.java]: I suppose these changes comes from previous commits

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedAdnITTDifferenceDateCalculation.java [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+            + "                            AND obs.value_datetime <= :previousPeriod "
+            + "                       GROUP  BY pa.patient_id "
+            + "                   ) most_recent "
+            + "               GROUP BY most_recent.patient_id "
+            + "               HAVING final_encounter_date < :previousPeriod ";
+
+    StringSubstitutor stringSubstitutor1 = new StringSubstitutor(map1);
+    String replacedQeury1 = stringSubstitutor1.replace(query1);
+
+    String finalQuery = replacedQeury1;
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedAdnITTDifferenceDateCalculation.java]: You can just initialize this variable with stringSubstitutor1.replace(query1); directly, to remove variable redundacny

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java [hunk 6]
```diff
@@ -191,25 +198,185 @@ private CalculationResultMap getMostRecentDateFromPatientWhoExperincedITT(
...
+    map2.put("returnVisitDateConcept", commonMetadata.getReturnVisitDateConcept().getConceptId());
+    map2.put(
+        "returnVisitDateForArvDrugConcept",
+        hivMetadata.getReturnVisitDateForArvDrugConcept().getConceptId());
+    map2.put("artDatePickup", hivMetadata.getArtDatePickupMasterCard().getConceptId());
+
+    String query2 =
+        " SELECT pat.patient_id,  eeecc.encounter_datetime   "
+            + " FROM   patient pat "
+            + " INNER JOIN encounter eeecc ON eeecc.encounter_id = pat.patient_id "
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java]: When having table aliases, can we have something that will be closer to the object? for example renaming **eeecc** to **enc** maybe the first 3 latters of a relation can be used and if there could be several, we could differentiate them with numbers in addition to letters

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java [hunk 7]
```diff
@@ -211,7 +279,11 @@ private CalculationResultMap getMostRecentDateFromPatientWhoExperincedITT(
     StringSubstitutor stringSubstitutor1 = new StringSubstitutor(map1);
     String replacedQeury1 = stringSubstitutor1.replace(query1);
 
-    //
+    return replacedQeury1;
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java]: you can just return **stringSubstitutor1.replace(query1);**

**Review by ningosi — COMMENTED** [review]
Hi @cfaife I might have repeated what is already sorted out, pardon me, please check through the comments again before we merge this for testing. Did you attempt tpo have a test for the calculation you wrote, it is huge!

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/EriDSDCohortQueries.java (hunk 3 above)
Comment (cfaife) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/EriDSDCohortQueries.java]: I guess it changed by `mvn clean install`

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedAdnITTDifferenceDateCalculation.java (hunk 5 above)
Comment (cfaife) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedAdnITTDifferenceDateCalculation.java]: done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java (hunk 6 above)
Comment (cfaife) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/rtt/ReturnedDateIITDateDaysCalculation.java]: done

**Review by ningosi — APPROVED** [review]
This is approved. @cfaife remember to have tests for your newly introduced calculation

**Comment by cfaife** [comment]
Ok thanks

**Comment by cfaife** [comment]
Are you going to merge now?

**Comment by ningosi** [comment]
Yes. Just waiting for it to finish running on the server

**Pull request merged**